File: qbs_ultra_lube/qbs_ultra_plus_lubes_pvt_ltd/report/sample_reg/sample_reg8jan.js
// Copyright (c) 2025, Astha and contributors
// For license information, please see license.txt



frappe.query_reports["Sample Reg"] = {
    "filters": [
{
    "fieldname": "customer",
    "label": __("Customer"),
    "fieldtype": "MultiSelectList",
    "options": "Customer",
    "get_data": function(txt) {
        return frappe.db.get_link_options("Customer", txt);
    }
},

        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date"
        },
        {
    "fieldname": "type_of_sample",
    "label": __("Sample Type"),
    "fieldtype": "Link",
    "options": "Sample Types",
    
}

    ],

    // onload(report) {
    //     // Reset dates on every load  
    //     report.set_filter_value("from_date", null);
    //     report.set_filter_value("to_date", null);

    //     // Add button  
    //     report.page.add_inner_button(__("Sample Registration Report"), function() {
    //         const filters = report.get_values();
    //         frappe.call({
    //             method: "rentals.rentals.report.sample_registration.sample_registration.execute",
    //             args: { filters },
    //             callback: function(r) {
    //                 const data = r.message && r.message[1];
    //                 if (data) {
    //                     open_custom_report_window(data);
    //                 } else {
    //                     frappe.msgprint(__("Could not fetch report data."));
    //                 }
    //             }
    //         });
    //     });
    // }
};


function open_custom_report_window(data) {
    if (!data || data.length === 0) {
        frappe.msgprint(__("No data to display."));
        return;
    }

    const pivoted_data = {};
    const sample_types = new Set();
    
    data.forEach(row => {
        if (row.type_of_sample) {
            sample_types.add(row.type_of_sample);
        }
        const customer = row.customer_name;
        
        if (!pivoted_data[customer]) {
            pivoted_data[customer] = { "customer_name": customer };
        }
        
        const field_prefix = row.type_of_sample.toLowerCase().replace(/ /g, '_').replace(/\./g, '');
        pivoted_data[customer][`${field_prefix}_td`] = row.created_today;
        pivoted_data[customer][`${field_prefix}_mtd`] = row.created_this_month;
        pivoted_data[customer][`${field_prefix}_ytd`] = row.created_this_year;
    });

    const final_data = Object.values(pivoted_data);
    const sorted_sample_types = Array.from(sample_types).sort();

    const tableHTML = `
        <!DOCTYPE html>
        <html>
        <head>
            <title>Sample Summary Report</title>
            <style>
                body { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif; padding: 20px; }
                table { width: 100%; border-collapse: collapse; font-size: 13px; }
                th, td { border: 1px solid #d1d8dd; padding: 8px; text-align: center; }
                th { background-color: #f5f7fa; font-weight: bold; }
                .print-button { padding: 8px 16px; background-color: #2490ef; color: white; border: none; border-radius: 4px; cursor: pointer; margin-bottom: 20px; }
                @media print { .print-button { display: none; } }
            </style>
        </head>
        <body>
            <h2>Sample Summary Report</h2>
            <table>
                <thead>
                    <tr>
                        <th rowspan="2" style="vertical-align: middle;">Name of Customer</th>
                        ${sorted_sample_types.map(st => `<th colspan="3">${st}</th>`).join('')}
                    </tr>
                    <tr>
                        ${sorted_sample_types.map(() => `<th>TD</th><th>MTD</th><th>YTD</th>`).join('')}
                    </tr>
                </thead>
                <tbody>
                    ${final_data.map(row => `
                        <tr>
                            <td>${row.customer_name || 'N/A'}</td>
                            ${sorted_sample_types.map(st => {
                                const field_prefix = st.toLowerCase().replace(/ /g, '_').replace(/\./g, '');
                                const td = row[`${field_prefix}_td`] || 0;
                                const mtd = row[`${field_prefix}_mtd`] || 0;
                                const ytd = row[`${field_prefix}_ytd`] || 0;
                                return `<td>${td}</td><td>${mtd}</td><td>${ytd}</td>`;
                            }).join('')}
                        </tr>
                    `).join('')}
                </tbody>
            </table>
        </body>
        </html>
    `;

    const newWindow = window.open('', '_blank');
    newWindow.document.write(tableHTML);
    newWindow.document.close();
}

File: qbs_ultra_lube/qbs_ultra_plus_lubes_pvt_ltd/report/sample_reg/sample_reg8jan.py
import frappe
from frappe import _
import json

@frappe.whitelist()
@frappe.whitelist()
def execute(filters=None):
    if filters and isinstance(filters, str):
        try:
            filters = json.loads(filters)
        except (json.JSONDecodeError, TypeError):
            filters = {}

    columns = get_columns()
    conditions, values = get_conditions(filters)
    data = get_data(conditions, values)

    if data:
        total_td = sum(row.get("created_today", 0) for row in data)
        total_mtd = sum(row.get("created_this_month", 0) for row in data)
        total_ytd = sum(row.get("created_this_year", 0) for row in data)

        total_row = {
            "customer_name": "Total",
            "type_of_sample": "",
            "created_today": total_td,
            "created_this_month": total_mtd,
            "created_this_year": total_ytd
        }

        data.append(total_row)

    chart = get_chart_data(data, filters)

    return columns, data, None, chart



def get_columns():
    return [
        { "label": _("Customer Name"), "fieldname": "customer_name", "fieldtype": "Link", "options": "Customer", "width": 250 },
        { "label": _("Sample Type"), "fieldname": "type_of_sample", "fieldtype": "Data", "width": 200 },
        { "label": _("TD"), "fieldname": "created_today", "fieldtype": "Int", "width": 180 },
        { "label": _("MTD"), "fieldname": "created_this_month", "fieldtype": "Int", "width": 200 },
        { "label": _("YTD"), "fieldname": "created_this_year", "fieldtype": "Int", "width": 180 }
    ]


def get_conditions(filters):
    conditions = ["docstatus = 1"]
    values = {}

    if not filters:
        return conditions, values


    if filters.get("customer"):
        customer_list = filters.get("customer")

        if isinstance(customer_list, str):
            customer_list = [
                c.strip()
                for c in customer_list.split(",")
                if c.strip()
            ]

        conditions.append("name_of_customer IN %(customer)s")
        values["customer"] = customer_list

    if filters.get("type_of_sample"):
        conditions.append("type_of_sample = %(type_of_sample)s")
        values["type_of_sample"] = filters.get("type_of_sample")

    if filters.get("from_date"):
        conditions.append("DATE(date_of_sample__receipt) >= %(from_date)s")
        values["from_date"] = filters.get("from_date")
        
    if filters.get("to_date"):
        conditions.append("DATE(date_of_sample__receipt) <= %(to_date)s")
        values["to_date"] = filters.get("to_date")

    return conditions, values


def get_data(conditions, values):
    ref_date = values.get("to_date") or frappe.utils.today()
    values["reference_date"] = ref_date
    
    sql_query = f"""
        SELECT
            name_of_customer AS customer_name,
            type_of_sample,
            SUM(CASE WHEN DATE(date_of_sample__receipt) = DATE(%(reference_date)s) THEN 1 ELSE 0 END) AS created_today,
            SUM(CASE WHEN YEAR(date_of_sample__receipt) = YEAR(%(reference_date)s)
                     AND MONTH(date_of_sample__receipt) = MONTH(%(reference_date)s)
                THEN 1 ELSE 0 END) AS created_this_month,
            SUM(CASE WHEN YEAR(date_of_sample__receipt) = YEAR(%(reference_date)s)
                THEN 1 ELSE 0 END) AS created_this_year
        FROM `tabSample Registration`
        WHERE {" AND ".join(conditions)}
        GROUP BY name_of_customer, type_of_sample
        ORDER BY name_of_customer, type_of_sample
    """
    return frappe.db.sql(sql_query, values, as_dict=True)


def get_chart_data(data, filters=None):
    if not data:
        return None
    data = [row for row in data if row.get("customer_name") != "Total"]
    customers = sorted(
        list(set(d['customer_name'] for d in data if d.get('customer_name')))
    )
    sample_types = sorted(
        list(set(d['type_of_sample'] for d in data if d.get('type_of_sample')))
    )

    pivot_data = {}
    for row in data:
        customer = row['customer_name']
        sample_type = row['type_of_sample']

        if not customer or not sample_type:
            continue

        if customer not in pivot_data:
            pivot_data[customer] = {}

        pivot_data[customer][sample_type] = row.get('created_this_year', 0)

    datasets = []
    for st in sample_types:
        values = [pivot_data.get(cust, {}).get(st, 0) for cust in customers]
        datasets.append({
            "name": st,
            "values": values
        })

    return {
        "data": {
            "labels": customers,
            "datasets": datasets
        },
        "type": "bar",
        "stacked": 1,
        "height": 900
    }